File: 6502.ml


type opcode = ADC | AND | ASL | BCC | BCS | BEQ | BIT | BMI | BNE | BPL | BRK | BVC | 
    BVS | CLC | CLD | CLI | CLV | CMP | CPX | CPY | DEC | DEX | DEY | EOR | INC | INX | 
    INY | JMP | JSR | LDA | LDX | LDY | LSR | NOP | ORA | PHA | PHP | PLA | PLP | ROL | 
    ROR | RTI | RTS | SBC | SEC | SED | SEI | STA | STX | STY | TAX | TAY | TSX | TXA | 
    TXS | TYA |
    (* undefined / invalid / undocumented TODO: http://nesdev.parodius.com/undocumented_opcodes.txt *)
    U__;;

type addrMode = Imm | Zer | Ixz | Iyz | Abs | Inx | Iny | Pre | Pst | Imp | Acc | Ind | Rel;;

(* http://nesdev.parodius.com/6502.txt *)
let opcodeMap = [|
(* Indexed by opcode, value is (mneumonic, addressing mode code) *)
(* x0         x1         x2         x3         x4        x5          x6         x7   *)
(* x8         x9         xa         xb         xc        xd          xe         xf   *)
(BRK, Imp);(ORA, Pre);(U__, Imm);(U__, Imm);(U__, Imm);(ORA, Zer);(ASL, Zer);(U__, Imm); (* 0x *)
(PHP, Imp);(ORA, Imm);(ASL, Acc);(U__, Imm);(U__, Imm);(ORA, Abs);(ASL, Abs);(U__, Imm); 
(BPL, Imp);(ORA, Pst);(U__, Imm);(U__, Imm);(U__, Imm);(ORA, Ixz);(ASL, Ixz);(U__, Imm); (* 1x *)
(CLC, Imp);(ORA, Iny);(U__, Imm);(U__, Imm);(U__, Imm);(U__, Imm);(ASL, Inx);(U__, Imm); 
(JSR, Abs);(AND, Pre);(U__, Imm);(U__, Imm);(BIT, Zer);(AND, Zer);(ROL, Zer);(U__, Imm); (* 2x *)
(PLP, Imp);(AND, Imm);(ROL, Acc);(U__, Imm);(BIT, Abs);(AND, Abs);(ROL, Abs);(U__, Imm); 
(BMI, Rel);(AND, Pst);(U__, Imm);(U__, Imm);(U__, Imm);(AND, Ixz);(ROL, Ixz);(U__, Imm); (* 3x *)
(SEC, Imp);(AND, Iny);(U__, Imm);(U__, Imm);(U__, Imm);(AND, Inx);(ROL, Inx);(U__, Imm); 
(EOR, Abs);(EOR, Pre);(U__, Imm);(U__, Imm);(U__, Imm);(EOR, Zer);(LSR, Zer);(U__, Imm); (* 4x *)
(PHA, Imp);(EOR, Imm);(LSR, Acc);(U__, Imm);(JMP, Abs);(RTI, Imp);(LSR, Abs);(U__, Imm);
(EOR, Inx);(EOR, Pst);(U__, Imm);(U__, Imm);(U__, Imm);(EOR, Ixz);(LSR, Ixz);(U__, Imm); (* 5x *)
(CLI, Imp);(EOR, Iny);(U__, Imm);(U__, Imm);(U__, Imm);(U__, Imm);(LSR, Inx);(U__, Imm);
(RTS, Imp);(ADC, Pre);(U__, Imm);(U__, Imm);(U__, Imm);(ADC, Zer);(ROR, Zer);(U__, Imm); (* 6x *)
(PLA, Imp);(ADC, Imm);(ROR, Acc);(U__, Imm);(JMP, Ind);(U__, Imm);(ROR, Abs);(U__, Imm);
(BVS, Rel);(ADC, Pst);(U__, Imm);(U__, Imm);(U__, Imm);(ADC, Ixz);(ROR, Ixz);(U__, Imm); (* 7x *)
(SEI, Imp);(ADC, Iny);(U__, Imm);(U__, Imm);(U__, Imm);(U__, Imm);(ROR, Inx);(U__, Imm);
(STA, Abs);(STA, Pre);(U__, Imm);(U__, Imm);(STY, Zer);(STA, Zer);(STX, Zer);(U__, Imm); (* 8x *)
(DEY, Imp);(U__, Imm);(TXA, Imp);(U__, Imm);(STY, Abs);(U__, Imm);(STX, Abs);(U__, Imm);
(STA, Inx);(STA, Pst);(U__, Imm);(U__, Imm);(STY, Ixz);(STA, Ixz);(STX, Ixz);(U__, Imm); (* 9x *)
(TYA, Imp);(STA, Iny);(TXS, Imp);(U__, Imm);(U__, Imm);(U__, Imm);(U__, Imm);(U__, Imm);
(LDY, Imm);(LDA, Pre);(LDX, Imm);(U__, Imm);(LDY, Zer);(LDA, Zer);(LDX, Zer);(U__, Imm); (* ax *)
(TAY, Imp);(LDA, Imm);(TAX, Imp);(U__, Imm);(LDY, Abs);(LDA, Abs);(LDX, Abs);(U__, Imm);
(BCS, Rel);(LDA, Pst);(U__, Imm);(U__, Imm);(LDY, Ixz);(LDA, Ixz);(LDX, Iyz);(U__, Imm); (* bx *)
(CLV, Imp);(LDA, Iny);(TSX, Imp);(U__, Imm);(LDY, Inx);(LDA, Inx);(LDX, Iny);(U__, Imm);
(CPY, Imm);(CMP, Pre);(U__, Imm);(U__, Imm);(CPY, Zer);(CMP, Zer);(DEC, Zer);(U__, Imm); (* cx *)
(INY, Imp);(CMP, Imm);(DEX, Imp);(U__, Imm);(CPY, Ixz);(CMP, Abs);(DEC, Abs);(U__, Imm);
(BNE, Rel);(CMP, Pst);(U__, Imm);(U__, Imm);(U__, Imm);(CMP, Ixz);(DEC, Ixz);(U__, Imm); (* dx *)
(CLD, Imp);(CMP, Iny);(U__, Imm);(U__, Imm);(U__, Imm);(CMP, Inx);(DEC, Inx);(U__, Imm);
(CPX, Imm);(SBC, Pre);(U__, Imm);(U__, Imm);(CPX, Zer);(SBC, Zer);(INC, Zer);(U__, Imm); (* ex *)
(INX, Imp);(SBC, Imm);(NOP, Imp);(U__, Imm);(CPX, Ixz);(SBC, Abs);(INC, Abs);(U__, Imm);
(BEQ, Rel);(SBC, Pst);(U__, Imm);(U__, Imm);(U__, Imm);(SBC, Ixz);(INC, Ixz);(U__, Imm); (* fx *)
(SED, Imp);(SBC, Iny);(U__, Imm);(U__, Imm);(U__, Imm);(SBC, Inx);(INC, Inx);(U__, Imm);
|];;

(* Bytes after opcode which operand requires for each addressing mode *)
let operandBytesForMode mode =
    match mode with
    | Imm -> 1
    | Zer -> 1
    | Ixz -> 1
    | Iyz -> 1
    | Abs -> 2
    | Inx -> 2
    | Iny -> 2
    | Pre -> 1
    | Pst -> 1
    | Imp -> 0
    | Acc -> 0
    | Ind -> 2
    | Rel -> 1;;

let nameOfMode mode =
    match mode with
    | Imm -> "Immediate"
    | Zer -> "Zero Page"
    | Ixz -> "Indexed X Zero Page"
    | Iyz -> "Indexed Y Zero Page"
    | Abs -> "Absolute"
    | Inx -> "Indexed X"
    | Iny -> "Indexed Y"
    | Pre -> "Pre-indexed Indirect"
    | Pst -> "Post-indexed Indirect"
    | Imp -> "Implied"
    | Acc -> "Accumulator"
    | Ind -> "Indirect"
    | Rel -> "Relative";;

(* TODO: formatters for each mode *)
(*
let addressingModes = [|
    (* (* index *) operand bytes, name, TODO: formatter *)
    (* Imm *) (1, "Immediate");               (* sprintf '#$%.IxzX',$_[Imm]} *)
    (* Zer *) (1, "Zero Page");               (* sprintf '$%.IxzX',$_[Imm]} *)
    (* Ixz *) (1, "Indexed X Zero Page");     (* sprintf '$%.IxzX,X',$_[Imm]} *)
    (* Iyz *) (1, "Indexed Y Zero Page");     (* sprintf '$%.IxzX,Y',$_[Imm]} *)
    (* Abs *) (2, "Absolute");                (* sprintf '$%.AbsX',($_[Imm])+(($_[Zer])*ImmxAccImm)} *)
    (* Inx *) (2, "Indexed X");               (* sprintf '$%.AbsX,X',($_[Imm])+(($_[Zer])*ImmxAccImm)} *)
    (* Iny *) (2, "Indexed Y");               (* sprintf '$%.AbsX,Y',($_[Imm])+(($_[Zer])*ImmxAccImm)} *)
    (* Pre *) (1, "Pre-indexed Indirect");    (* sprintf '($%.IxzX,X)', $_[Imm] } *)
    (* Pst *) (1, "Post-indexed indirect");   (* sprintf '($%.IxzX),Y', $_[Imm] } *)
    (* Imp *) (0, "Implied");                 (* '' *)
    (* Acc *) (0, "Accumulator");             (* 'A' *)
    (* Ind *) (2, "Indirect");                (* sprintf '($%.AbsX)', ($_[Imm])+(($_[Zer])*ImmxAccImm)} *)  (* JMP only *)
    (* Rel *) (1, "Relative");                (* sprintf '$%.AbsX', sign_num($_[Imm])+$_[Zer] } *) 
|];;
*)

File: go/6502.go
// Created:20101126
// By Jeff Connelly

//

package main

import "fmt"

// TODO: should these map to the opcodes themselves instead of opaque integers?? Probably.
type Opcode int 
const (___ = 0;
ADC; AND; ASL; BCS; BEQ; BIT; BMI; BNE; BRK; BVS; CLC; CLD; CLI; CLV; CMP; CPX; 
CPY; DEC; DEX; DEY; EOR; INC; INX; INY; JMP; JSR; LDA; LDX; LDY; LSR; NOP; ORA; 
PHA; PHP; PLA; PLP; ROL; ROR; RTI; RTS; SBC; SEC; SED; SEI; STA; STX; STY; TAX; 
TAY; TSX; TXA; TXS; TYA)

// TODO: AddrMode

// Opcode and addressing mode
type OpcodeAddrMode struct { opcode, addrmode int }

// Indexed by opcode number, maps to decoded opcode and addressing mode
var opcodes = [...]OpcodeAddrMode{
// ==0==     ==1==     ==2==    ==3==     ==4==     ==5==     ==6==     ==7==
{BRK,  9},{ORA,  7},{___,  0},{___,  0},{___,  0},{ORA,  1},{ASL,  1},{___,  0},
{PHP,  9},{ORA,  0},{ASL, 10},{___,  0},{___,  0},{ORA,  4},{ASL,  4},{___,  0},
{ORA,  5},{ORA,  8},{___,  0},{___,  0},{___,  0},{ORA,  2},{ASL,  2},{___,  0},
{CLC,  9},{ORA,  6},{___,  0},{___,  0},{___,  0},{___,  0},{ASL,  5},{___,  0},
{JSR,  4},{AND,  7},{___,  0},{___,  0},{BIT,  1},{AND,  1},{ROL,  1},{___,  0},
{PLP,  9},{AND,  0},{ROL, 10},{___,  0},{BIT,  4},{AND,  4},{ROL,  4},{___,  0},
{BMI, 12},{AND,  8},{___,  0},{___,  0},{___,  0},{AND,  2},{ROL,  2},{___,  0},
{SEC,  9},{AND,  6},{___,  0},{___,  0},{___,  0},{AND,  5},{ROL,  5},{___,  0},
{EOR,  4},{EOR,  7},{___,  0},{___,  0},{___,  0},{EOR,  1},{LSR,  1},{___,  0},
{PHA,  9},{EOR,  0},{LSR, 10},{___,  0},{JMP,  4},{RTI,  9},{LSR,  4},{___,  0},
{EOR,  5},{EOR,  8},{___,  0},{___,  0},{___,  0},{EOR,  2},{LSR,  2},{___,  0},
{CLI,  9},{EOR,  6},{___,  0},{___,  0},{___,  0},{___,  0},{LSR,  5},{___,  0},
{RTS,  9},{ADC,  7},{___,  0},{___,  0},{___,  0},{ADC,  1},{ROR,  1},{___,  0},
{PLA,  9},{ADC,  0},{ROR, 10},{___,  0},{JMP, 12},{___,  0},{ROR,  4},{___,  0},
{BVS, 12},{ADC,  8},{___,  0},{___,  0},{___,  0},{ADC,  2},{ROR,  2},{___,  0},
{SEI,  9},{ADC,  6},{___,  0},{___,  0},{___,  0},{___,  0},{ROR,  5},{___,  0},
{STA,  4},{STA,  7},{___,  0},{___,  0},{STY,  1},{STA,  1},{STX,  1},{___,  0},
{DEY,  9},{___,  0},{TXA,  9},{___,  0},{STY,  4},{___,  0},{STX,  4},{___,  0},
{STA,  5},{STA,  8},{___,  0},{___,  0},{STY,  2},{STA,  2},{STX,  2},{___,  0},
{TYA,  9},{STA,  6},{TXS,  9},{___,  0},{___,  0},{___,  0},{___,  0},{___,  0},
{LDY,  0},{LDA,  7},{LDX,  0},{___,  0},{LDY,  1},{LDA,  1},{LDX,  1},{___,  0},
{TAY,  9},{LDA,  0},{TAX,  9},{___,  0},{LDY,  4},{LDA,  4},{LDX,  4},{___,  0},
{BCS, 12},{LDA,  8},{___,  0},{___,  0},{LDY,  2},{LDA,  2},{LDX,  3},{___,  0},
{CLV,  9},{LDA,  6},{TSX,  9},{___,  0},{LDY,  5},{LDA,  5},{LDX,  6},{___,  0},
{CPY,  0},{CMP,  7},{___,  0},{___,  0},{CPY,  1},{CMP,  1},{DEC,  1},{___,  0},
{INY,  9},{CMP,  0},{DEX,  9},{___,  0},{CPY,  2},{CMP,  4},{DEC,  4},{___,  0},
{BNE, 12},{CMP,  8},{___,  0},{___,  0},{___,  0},{CMP,  2},{DEC,  2},{___,  0},
{CLD,  9},{CMP,  6},{___,  0},{___,  0},{___,  0},{CMP,  5},{DEC,  5},{___,  0},
{CPX,  0},{SBC,  7},{___,  0},{___,  0},{CPX,  1},{SBC,  1},{INC,  1},{___,  0},
{INX,  9},{SBC,  0},{NOP,  9},{___,  0},{CPX,  2},{SBC,  4},{INC,  4},{___,  0},
{BEQ, 12},{SBC,  8},{___,  0},{___,  0},{___,  0},{SBC,  2},{INC,  2},{___,  0},
{SED,  9},{SBC,  6},{___,  0},{___,  0},{___,  0},{SBC,  5},{INC,  5},{___,  0},
}

func main() {
    fmt.Printf("Hello, world!\n")
}

